### PR TITLE
Improve cancellation, fix synchronization

### LINKED ISSUE #5: Add a `canceled` flag to tasks when execution is canceled by context cancellation
On thing could be improved: the error message on the task that is canceled through `Cancel` which uses cancellation through the `context.Context` is `context canceled`. It would be nice to add a `canceled` flag to tasks and show that instead in the UI.

_Originally posted by @hlubek in https://github.com/Flowpack/prunner/issues/4#issuecomment-891127311_

### PR DESCRIPTION
**Pull request by hlubek**
Fixes #5

### CODE REVIEW

File: prunner.go [hunk 1]
```diff
@@ -665,31 +676,28 @@ func (r *PipelineRunner) requestPersist() {
...
+		WithField("pipeline", job.Pipeline).
+		WithField("jobID", job.ID).
+		Debugf("Canceling job")
 
-	// SAFEGUARD: it could happen that a job is cancelled which has never been scheduled.
-	// thus, we need to check for the existence of the job scheduler before cancelling.
-	if job.sched != nil {
-		job.sched.Cancel()
-	}
+	go job.sched.Cancel()
```
Comment (hlubek) [inline prunner.go:700]: This should do the trick of making sure `job.sched` is not nil and keep the go routine from deadlocking on status updates while calling `Cancel()`.

File: prunner.go [hunk 2]
```diff
@@ -362,14 +363,15 @@ func (r *PipelineRunner) HandleTaskChange(t *task.Task) {
...
-	jt.Errored = t.Errored
-	jt.Error = t.Error
 	jt.ExitCode = t.ExitCode
 	jt.Skipped = t.Skipped
 
 	// Set canceled flag on the job if a task was canceled through the context
 	if errors.Is(t.Error, context.Canceled) {
-		j.Canceled = true
+		jt.Canceled = true
+	} else {
```
Comment (hlubek) [inline prunner.go:372]: We shouldn't show tasks that were canceled as having an error, so we just set the `Canceled` flag in this case.

File: prunner.go [hunk 3]
```diff
@@ -392,7 +394,11 @@ func (r *PipelineRunner) HandleStageChange(stage *scheduler.Stage) {
 		return
 	}
 
-	jt.Status = toStatus(stage.ReadStatus())
+	if jt.Canceled {
```
Comment (hlubek) [inline prunner.go:397]: Here the task flag is "promoted" to the task status which is set by stage updates from the scheduler.

**Review by skurfuerst — APPROVED** [review]
looks good, let's try this in prod :)

**Pull request merged**